homework-05/src/main.rs

```diff
@@ -0,0 +1,172 @@
+        return Err("Input contains non-alphabetic characters".into());
+    }
+    Ok(input.to_lowercase())
+}
+
+fn uppercase(input: &str) -> Result<String, Box<dyn Error>> {
+    Ok(input.to_uppercase())
+}
+
+fn alphabetic_uppercase(input: &str) -> Result<String, Box<dyn Error>> {
+    if input.chars().any(|c| !c.is_alphabetic()) {
+        return Err("Input contains non-alphabetic characters".into());
+    }
+    Ok(input.to_lowercase())
+}
+
+fn no_spaces(input: &str) -> Result<String, Box<dyn Error>> {
+    Ok(input.replace(" ", ""))
+}
+
+fn slugify_text(input: &str) -> Result<String, Box<dyn Error>> {
+    Ok(slugify(input))
+}
+
+fn kebab(input: &str) -> Result<String, Box<dyn Error>> {
+    Ok(input.replace(" ", "-").to_lowercase())
+}
+
+fn snake(input: &str) -> Result<String, Box<dyn Error>> {
+    Ok(input.replace(" ", "_").to_lowercase())
+}
+
+fn hex(input: &str) -> Result<String, Box<dyn Error>> {
+    Ok(input.chars().map(|c| format!("{:02x}", c as u8)).collect())
+}
+
+fn get_column_widths(headers: &csv::StringRecord, records: &[csv::StringRecord]) -> Vec<usize> {
+    let mut widths: Vec<usize> = headers.iter().map(|h| h.len()).collect();
+    for record in records {
+        for (i, field) in record.iter().enumerate() {
+            if field.len() > widths[i] {
+                widths[i] = field.len();
+            }
+        }
+    }
+    widths
+}
+
+fn format_csv_separator_line(widths: &[usize]) -> String {
+    widths.iter().map(|&w| format!("+{}", "-".repeat(w + 2))).collect::<Vec<_>>().join("") + "+\n"
+}
+
+fn format_csv_line(fields: &[&str], widths: &[usize]) -> String {
+    fields.iter().enumerate().map(|(i, field)| format!("| {:<width$} ", field, width = widths[i])).collect::<Vec<_>>().join("") + "|\n"
+}
+
+fn csv(input: &str) -> Result<String, Box<dyn Error>> {
+    let mut reader = Reader::from_reader(input.as_bytes());
+    let headers = reader.headers()?.clone();
+    let records = reader.records().collect::<Result<Vec<_>, _>>()?;
+
+    // Check if all records have the same number of columns as the headers
+    for record in &records {
+        if record.len() != headers.len() {
+            return Err(format!("Data line does not conform to header line: expected {} columns, found {} columns", headers.len(), record.len()).into());
+        }
+    }
+
+    // Get the width of each column
+    let widths = get_column_widths(&headers, &records);
+
+    // Create a separator line
+    let separator = format_csv_separator_line(&widths);
+
+    // Create the header line
+    let header_line = format_csv_line(&headers.iter().collect::<Vec<_>>(), &widths);
+
+    // Create the value lines
+    let value_lines: String = records.iter().map(|record| {
+        format_csv_line(&record.iter().collect::<Vec<_>>(), &widths)
+    }).collect::<Vec<_>>().join("");
+
+    // Combine everything into the final output
+    let result = format!("{}{}{}{}", separator, header_line, separator, value_lines) + &separator;
```

Comment:
You could probably have the second separator in the format as well (you could put a {0} at the end to specifically request the first arg to the format again), and avoid the creation of an extra String :)